crates/polars-plan/src/plans/conversion/dsl_to_ir.rs

```diff
@@ -459,6 +462,24 @@ pub fn to_alp_impl(
             convert.fill_scratch(&left_on, expr_arena);
             convert.fill_scratch(&right_on, expr_arena);
 
+            let all_elementwise = |aexprs: &[ExprIR]| {
+                // Single non-elementwise key is fine, it can't
+                // mismatch with the length of the other ones (since
+                // they don't exist).
+                aexprs.len() <= 1
+                    // Otherwise every expression must be elementwise
+                    // so that we are guaranteed the keys for a join
+                    // are all the same length.
+                    || aexprs.iter().all(|expr_ir| {
```

Comment:
I think we can use `has_aexpr` here. 

I do think we should be stricter here. For instance a `head` is not allowed as its length doesn't match the dataframe. Probably allowing only elementwise makes most sense, as it doesn't seem sensible to join by a `col(..).sort()` anyway.

Reply:
> I do think we should be stricter here. 

Should I drop the `aexprs.len() <= 1` part of the condition?

Is there a full classification of "elementwise" yet? As you note, `sort` answers "yes" to `single_aexpr_is_elementwise`, but should probably (?) be rejected.

Reply:
I think we can remove `single_aexpr_is_elementwise` and replace it with `is_streamable`. It seems redundant. `is_streamable` is more correctly implemented as well.

> Should I drop the aexprs.len() <= 1 part of the condition?

Yes.

Reply:
Done.